util/gpool/spmc/spmcpool.go

```diff
@@ -0,0 +1,403 @@
+	p.lock.Unlock()
+	// There might be some callers waiting in retrieveWorker(), so we need to wake them up to prevent
+	// those callers blocking infinitely.
+	p.cond.Broadcast()
+}
+
+func isClose(exitCh chan struct{}) bool {
+	select {
+	case <-exitCh:
+		return true
+	default:
+	}
+	return false
+}
+
+// ReleaseAndWait is like Release, it waits all workers to exit.
+func (p *Pool[T, U, C, CT, TF]) ReleaseAndWait() {
+	if p.IsClosed() || isClose(p.stopCh) {
+		return
+	}
+
+	close(p.stopCh)
+	p.release()
+	for {
+		if p.Running() == 0 && p.heartbeatDone.Load() {
+			return
+		}
+	}
+}
+
+// SetConsumerFunc is to set ConsumerFunc which is to process the task.
+func (p *Pool[T, U, C, CT, TF]) SetConsumerFunc(consumerFunc func(T, C, CT) U) {
+	p.consumerFunc = consumerFunc
+}
+
+// AddProduceBySlice is to add Produce by a slice.
+// Producer need to return ErrProducerClosed when to exit.
+func (p *Pool[T, U, C, CT, TF]) AddProduceBySlice(producer func() ([]T, error), constArg C, contextFn TF, options ...TaskOption) (<-chan U, pooltask.TaskController[T, U, C, CT, TF]) {
+	opt := loadTaskOptions(options...)
+	taskID := p.NewTaskID()
+	var wg sync.WaitGroup
+	result := make(chan U, opt.ResultChanLen)
+	closeCh := make(chan struct{})
+	tc := pooltask.NewTaskController[T, U, C, CT, TF](p, taskID, closeCh, &wg)
+	taskCh := make(chan pooltask.Task[T], opt.TaskChanLen)
+	for i := 0; i < opt.Concurrency; i++ {
+		err := p.run()
+		if err == gpool.ErrPoolClosed {
+			break
+		}
+		taskBox := pooltask.NewTaskBox[T, U, C, CT, TF](constArg, contextFn, &wg, taskCh, result, taskID)
+		p.taskCh <- &taskBox
+	}
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				logutil.BgLogger().Error("producer panic", zap.Any("recover", r), zap.Stack("stack"))
+			}
+			close(closeCh)
+			close(taskCh)
+		}()
+		for {
+			tasks, err := producer()
+			if err != nil {
+				if errors.Is(err, gpool.ErrProducerClosed) {
+					return
+				}
+				log.Error("producer error", zap.Error(err))
+				return
+			}
+			for _, task := range tasks {
+				wg.Add(1)
+				task := pooltask.Task[T]{
+					Task: task,
+				}
+				taskCh <- task
+			}
+		}
+	}()
+	return result, tc
+}
+
+// AddProducer is to add producer.
+// Producer need to return ErrProducerClosed when to exit.
+func (p *Pool[T, U, C, CT, TF]) AddProducer(producer func() (T, error), constArg C, contextFn TF, options ...TaskOption) (<-chan U, pooltask.TaskController[T, U, C, CT, TF]) {
+	opt := loadTaskOptions(options...)
+	taskID := p.NewTaskID()
+	var wg sync.WaitGroup
+	result := make(chan U, opt.ResultChanLen)
+	closeCh := make(chan struct{})
+	tc := pooltask.NewTaskController[T, U, C, CT, TF](p, taskID, closeCh, &wg)
+	taskCh := make(chan pooltask.Task[T], opt.TaskChanLen)
+	for i := 0; i < opt.Concurrency; i++ {
+		err := p.run()
+		if err == gpool.ErrPoolClosed {
+			break
+		}
+		taskBox := pooltask.NewTaskBox[T, U, C, CT, TF](constArg, contextFn, &wg, taskCh, result, taskID)
+		p.taskCh <- &taskBox
```

Comment:
Could we rename this field `p.taskCh`? Its name is the same `taskCh`(line276), but they are different types.

Reply:
I rename the ```taskCh``` in line276 with ```inputCh```.